### DISCUSSION TITLE
Read-Write support - Send command testing

### DISCUSSION THREAD

**Discussion by netmindz** [netmindz/balboa_GL_ML_spa_control#25] [General]
Please use this discussion for development and testing discussion on the current work in progress on actually controlling the tub, rather than just viewing its status 

https://github.com/netmindz/balboa_GL_ML_spa_control/pull/27

**Comment by tmjo** [comment]
Hi,

I did some preliminary testing yesterday, unfortunately no success, and ended up taking a hot bath instead of fiddling around more last night :) Did some brief testing of both branches and also tried to edit the constant.h file to fit my tub. I did some reading of the discussions in the old thread and have some questions though:

- You mention that you tested writing with 5V instead of 3.3V which made  difference. I've currently adjusted my PSU to 3.3V. Do I understand correct that you had to adjust everything to 5V to make writing work? If so, are you running your ESP on 5V too, or just the supply to RS485 board?

- What is the latest info we have on FB-command formatting. I see there are some minor differences between our Wiki and the discussions you had with TiSiAi. My interpretation after reading is that in
`FB06 66666666 09F6 88`
FB06 is an instruction to give command, 66666666 is an UID or something that has to the specific for each pool (could it possibly be the model of a panel?), the next 09F6 is the actual command and 88 is a crc code? Did I get it right? Or are the last four F688 the crc?

I see from my previous loggings that the temp up/down (my tub has only one button for this) is giving `fb06 03430600 05` without any crc at the end which seems confusing. The other commands gave a 'full blown' command like `fb06 03430600 09f6 4c`.

Looking forward to testing some more here. Any advice is welcome!

**Reply by netmindz** [reply]
ok, lets try and break that down a bit -

1) You mention two branches, not sure exactly which 2 you mean.

2) The FB commands with 66666666 will not work without editing as TiSiAi removed his own UID, but didn't then calculate the right CRC value so they definitely will not work. What I don't know is if you try and use my commands, which have a valid CRC will work or if they will be ignored due to the "unexpected" UID. You might need to try say swapping the toggle light to a previously captured FB value for that from your own setup. I would only try this however when you are sure mine don't work.

3) With regards to 3.3v vs 5v, the input power for the ESP32 is always 5v as it's USB. I had been running the RS485 adapter on the 3.3v output of the ESP32 as I was avoiding putting 5v output of the RS485 back into the ESP32, however, I couldn't get any response at all from the tub with this setup. So looks like you can read RS485 with only 3.3v but certainly with the adapter I'm using you can't write/drive the RS485 with only 3.3v

4) The fact your single-use temp button seems a bit weird, I'm unsure about it. That's why I would suggest testing with the toggle light command just to prove you can send *something* to the tub in the first instance

**Reply by netmindz** [reply]
It's perhaps worth saying that I was unable to get the delay between FA and FB that TiSiAi uses in his project to work for me, but then I think that might have been before I had established the level shifter delay and rs485 voltage issue, so for now I'm just resending the same command until I see the controller respond. 

I might try the single command, time based approach again at some point

**Comment by tmjo** [comment]
Perhaps I wasn't too clear :)

> ok, lets try and break that down a bit -
> 
> 1. You mention two branches, not sure exactly which 2 you mean.

Sorry, I meant the two control branches `send-command-queue` and `send-command-v2`

> 2. The FB commands with 66666666 will not work without editing as TiSiAi removed his own UID, but didn't then calculate the right CRC value so they definitely will not work. What I don't know is if you try and use my commands, which have a valid CRC will work or if they will be ignored due to the "unexpected" UID. You might need to try say swapping the toggle light to a previously captured FB value for that from your own setup. I would only try this however when you are sure mine don't work.

Yes, it was just an example. I tried the unmodified `constant.h` file before I remembered more or less how it works, and then changed to 
"my own" data from my previously captured data/logging. I'll do some more testing though.

> 3. With regards to 3.3v vs 5v, the input power for the ESP32 is always 5v as it's USB. I had been running the RS485 adapter on the 3.3v output of the ESP32 as I was avoiding putting 5v output of the RS485 back into the ESP32, however, I couldn't get any response at all from the tub with this setup. So looks like you can read RS485 with only 3.3v but certainly with the adapter I'm using you can't write/drive the RS485 with only 3.3v

Right. I don't have USB connected, but I guess the Vin on ESP is 5V anyway, so I can trim my PSU to 5V. Perhaps this is the reason why it didn't seem to write anything then.

> 4. The fact your single-use temp button seems a bit weird, I'm unsure about it. That's why I would suggest testing with the toggle light command just to prove you can send _something_ to the tub in the first instance

Agreed. I tried both light, mode and temp up/down. I am also hoping that I can use separate ups and downs instead of the 'toggle', it could even be possible we can get functionality corresponding to the most advanced controller even if our panels are not.

I'll tune up to 5V and see if I get some better response.

**Comment by tmjo** [comment]
Ding, ding, ding. 5V is my friend, I have control. I did not test everything yet, but I was able to switch on light as well as changing temperature up and down using your commands in constant.h (i.e. without modifying to my logged results). Looking awesome!

**Reply by netmindz** [reply]
The eco-only should flip between eco and standard, I use that with automation for when electricity is cheaper

**Reply by netmindz** [reply]
For the temp up and down, was that using the up and down buttons on the v2 branch? I wrote that but haven't tested yet, only used up and down to cycle when changing mode

**Reply by tmjo** [reply]
I will try it. Temp was by using the buttons for up and down i v2, yes. Seems to be working well! My temp setpoint has always been a bit buggy, so not 100% sure it was showing correct, but at least it moved both up and down :)

**Reply by tmjo** [reply]
So, tried a bit more. It's not working perfectly for me yet, but that would be as expected and we're very very close.

Adjusting the temperature with up/down: It kindof works, but it doesn't stop and counts to minimum and maximum temperatures. Seems stuck in a loop there and had to reboot the ESP to get back to normal as the panel seemed to be blocked and I was stuck in "menu 46" status and display flashing as if it was in temperature adjustment mode still. 

When trying to change mode, I can see the panel is reacting and starts flashing for standard mode. I was able to toggle the modes with temp up/down, but the mode doesn't activate unless I press light manually. Didn't study the code details yet, perhaps that is all that is implemented for now?

Anyway, this is so cool!

**Reply by netmindz** [reply]
The *only* known good stuff on v2 is the light, the rest is very much a work in progress as I try and build stuff "properly".

if you want something that I've actually tested and is my "production" version at the moment, use `send-command-queue`

**Reply by netmindz** [reply]
I'm tempted to remove the "raw" buttons as unless you can replicate the current status of the display, including flashing them you are kinda working blind.

Still work needed for the climate/hvac style setting of the target temp, but buggy as it is, it's still better than having to go outside and press the buttons. Not reliable enough to be set using with an automation in HA however yet

**Comment by netmindz** [comment]
It's not 100% reliable yet, but I've just pushed the "climate" set temp control now. The operation mode does nothing. Need to see why that's coming through despite not being an enabled feature
![Screenshot from 2023-06-20 16-37-24](https://github.com/netmindz/balboa_GL_ML_spa_control/assets/442066/1201a483-93b1-4954-bbc2-c02cae3ffa6f)

**Comment by netmindz** [comment]
Do you have a FB command for Jets/pump 1 and 2 @tmjo ?  Just seen the values I've put in constants are 1 char too short, so that's definitely not going to work

**Reply by tmjo** [reply]
From my notes I have:
```
	Jets		fb06 03430600 06f9 a2
	Aux		fb06 03430600 07f8 b0
```
The button on my panel says Aux, but it is pump2 I guess (my second pump is broken, so I've never used in in real life).

**Reply by netmindz** [reply]
I can now turn my pumps on - just need to finish the TODO so I can switch off :)

**Reply by netmindz** [reply]
https://docs.google.com/spreadsheets/d/1Awr4f3IpCDEJkqJy9oKkBBd2rB1e10lnuQLhyQ15IZ4/edit#gid=0 - this is for the mode switch, but the same issue for how many button presses to loop the options for pump speed if you fancy solving the riddle @tmjo ?

**Reply by tmjo** [reply]
I'll take a look at this. I did some testing on latest send-command-v2 branch, but it was during bathing, so no real good debugging available. 

Pumps came on, but not off and not high speed although I've set that option. There was a reaction, so it is probably just some minor adjustments. 

Modes and light seems to be working well, that's great. We should implement standby mode as well. 

For temperature up/down, my tub still goes crazy and goes to min/max respectively. It blocks my panel too, but it came back to life after a couple of minutes so I didn't have to reboot ESP this time. I'll take a look, taking into account I only have the toggle stuff and not real up/down buttons, there could be something particular for my setup.

Nice when you can enjoy some nerding on the phone while taking a relaxing bath and enjoying a cold beer :)

PS! I saw you mentioned being close to merging the branch into main; I vote for, although there is still things to do and to fix, I think it is close enough. Also we could probably do some refactoring to make code more efficient, perhaps a separate discussion thread for this issue?

Cheers!

**Reply by netmindz** [reply]
Ok, not entirely surprised that dual speed pumps didn't work on first attempt. If you can send the serial output I can see what it's trying to do.

Is standby mode just a different name for sleep or a mode in it's own right?

For temp change - are you trying to use the buttons or the set target temp widget? While you open top panel doesn't have independent buttons for temp, I wouldn't have thought the control unit would therefore struggle with a different top panel having them.

Now I've created a read-only branch that only needs the RS485 connection and not the risky 5v stuff, I think I'm happy to update the docs and merge down

**Comment by caldog20** [comment]
This may have already been covered, but between the different commands with different UUID's we have, I did find a pattern between all of them. 

The 2 command bytes in the command string are the inverse of the second to last byte before the crc byte. 

For instance, the following commands are in order:

`fb0603450e0000ff<crc>`
`fb0603450e0001fe<crc>`
`fb0603450e0003fc<crc>`

The pattern looks like this:

```
Command bytes:
0000 - ff - empty
0001 - fe
0002 - fd
0003 - fc - time
0004 - fb - change mode
0005 - fa
0006 - f9 - jet1
0007 - f8 - jet2
0008 - f7
0009 - f6 - light
```


Using this, you can use the crc generator and generate the last byte (crc byte) and the commands work. This may not be technically correct but it seems to be working. Just thought I would share a pattern I found.

**Reply by netmindz** [reply]
I had not spotted that pattern, nice find. 

TiSiAi provided a CRC function, but I'm unsure how to actually use it

**Reply by caldog20** [reply]
That was the CRC function I was playing around with. I actually am not sure if it's 100% correct though.

 I went back through the first discussion and found some of TiSiAi's captures. I took those values and validated them against the CRC function they provided and for every command using their panel ID, the CRC byte matched. However, for the commands that had different panel IDs (from `constants.h`), the crc byte the function returns did not match. Here is a link to what I was doing when I was testing with it.

https://godbolt.org/z/TEqhrhvPc

**Reply by netmindz** [reply]
It would be good to fill in the gaps there, we are missing blower for example

**Comment by caldog20** [comment]
Also I found that the time is displayed in the place of the temperature, but only after pressing the time button once (for an ML700 anyways)

I was able to successfully set the time and get the results back in HA by periodically sending a time button press to read out the time.

A time status message looks almost exactly like a normal temp status message. Here is a time status message for example:

`fa14303531360083120000088002ffffff0000000000db`

This shows its currently set for 05:16

**Reply by caldog20** [reply]
Since the time status message is similar to the temperature in the temp status message, the time needs to be parsed from hex to ascii then it can be sent to HA.

**Reply by netmindz** [reply]
We already have code that reads the time, even when it's not on screen. Technically the reading of the current temp is actually more reading what's on screen - kinda. 

@tmjo did find that the temp in F is actually returned elsewhere, see one of the PRs, we didn't bother merging so far as it didn't give greater temp accuracy than 0.5 shown on screen, but as part of the refactor cleanup of the code it might be better to flip to using that

If you have the ML700 it would be great to have the option to trigger setting the tub time to match NTP for those of use with more basic panels. I can't set the time of my cleaning cycle to match my cheap electricity other than by powering off the tub and clearing it's settings to reset where in a 24 hour cycle we are.

**Reply by caldog20** [reply]
Yeah I just couldn't get that to work. Maybe there is another setting that needs to be done but with the time set I still get back ffff in regular status messages. The only way I get get the time to come back is by emulating a time button press. 

There wasn't a time button yet so I was just adding it and messing around with it.

**Reply by caldog20** [reply]
Ok I see what you are talking about with the temp from F in the time status message. 

Here is a temp status message I got (basically when the topside panel screen is showing temp): `fa14203932460000220000008002ffff5c0000000000fa`. The `5c` is the current temp in F. 

As far as time messages (when topside panel is showing time), here is what I get: `fa14313034310040120000008002ffffff00000000002c`. Here `31303431` is my current time setting. 

I haven't figured out yet why I don't get the time back in the temp status message. I am thinking there may be some differences between the actual current time, and the time filter cycles are scheduled, or time til the next cycle?

I am going to start reading through this manual (where I learned how to set the time). I believe anyone could use the procedures in this manual to configure their board even if you don't have an advanced panel.

One important thing: If you are currently using a panel that is less advanced, it is possible that you have the jump connected on the main board that disables the Real-time Clock (RTC). For instance, to use the ML700 panel to program filter/cleaning times, you must have the RTC jumper disconnected so the clock on the board functions properly. 


Here is the manual I have been reading: https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwj-vt7I6ej_AhXDl2oFHU9MDEIQFnoECAIQAQ&url=https%3A%2F%2Fwww.balboawatergroup.com%2Fgetdoc.cfm%3Fid%3D1370&usg=AOvVaw2LfdZrCitWy_qnVy7W-9im&opi=89978449


One thing I want to do eventually is to start decoding error messages and implement them in the code. I believe a fair amount of them can be triggered manually (i.e. unplug temp sensor A, check messages, repeat for B). It would be nice to have diagnostic messages incase of an issue. Especially in my case where my ML700 panel is trash.

**Reply by netmindz** [reply]
Yeah at the moment I'm completely ignoring the AE 0D messages. At one point I did get different values, but mine are just static at the moment, so without being able to make a known change, I can't see how it affects the value.  I hadn't considered that they might include diagnostic/fault codes.

Yeah I'm familiar with the dip switch to set clock Vs time. Would be to change in order to test the setting of the time

**Reply by caldog20** [reply]
The RTC is a jumper and not a dipswitch on my board, but I think there is a dipswitch possibly for a time setting as well.

**Reply by netmindz** [reply]
A3
. . . . . . . . . . . . . . . . In “ON” position, filter cycles are programmed by duration for non-time capable panels
. . . . . . . . . . . . . . . . In “OFF” position, filter cycles are programmed to start and end times for time capable panels

**Reply by caldog20** [reply]
This was the jumper I was referring to. Probably different on my board.

![image](https://github.com/netmindz/balboa_GL_ML_spa_control/assets/10728984/fad04833-5820-43dd-9b1e-81f885872347)

**Reply by netmindz** [reply]
Looks like you might need to change both then

**Reply by netmindz** [reply]
Do you think we can generate the command sequence to set the tub to the current time which we've fetched from NTP @caldog20 ?

**Reply by caldog20** [reply]
> Do you think we can generate the command sequence to set the tub to the current time which we've fetched from NTP @caldog20 ?

Yeah for sure it's doable. Sorry I've been swamped with work lately. I'll look into it next week. I have a test to take this week to renew a certification.

**Reply by netmindz** [reply]
no worries, it just came to mind as a nice feature, especially for those of us without the more complex topside panels

**Reply by netmindz** [reply]
The latest release contains an "accidental feature" that at least for me with my 3 port controller is capturing the FB commands sent by my top panel as the FB value sent out on MQTT. 

Would be interested to know if you see the same @caldog20 as it would help capture the extra commands your ML700 can send without needing to swap back to older sketches that captured the full stream

**Comment by netmindz** [comment]
Just pushed an update, version 0.2.4 that adds a 500ms delay between commands and has better tracking of retry command sending. For multi-command actions like changing mode or temp it's a definite improvement.  I still want to remove the retry, but my own setup for some reason is still dropping some messages so do can't just yet. Hopefully the T-RSC3 will be more reliable about sending, but I'm using that for the ESPHome component dev at the moment

**Comment by netmindz** [comment]
I've pushed a tweak that optimises command sending a bit, still not perfect, but is improvement
